File: Book.js
class Book extends TableOfContents {
  constructor(title) {
    super();
    this.title = title;
    this.authors = [];
    this.chapters = [];
    this.tableOfContents = new TableOfContents();
  }

  addAuthor(authorOrName) {
    const author =
      typeof authorOrName === 'string' ? new Author(authorOrName) : authorOrName;
    this.authors.push(author);
    return author;
  }

  addChapter(name) {
    const chapter = new Chapter(name);
    this.chapters.push(chapter);
    this.tableOfContents.addEntry(name);
    return chapter;
  }

  getChapter(index) {
    return this.chapters[index];
  }

}

File: src/Book.java
import java.util.ArrayList;
import java.util.List;

public class Book extends Section {
    private final List<Author> authors = new ArrayList<>();
    private final TableOfContents tableOfContents = new TableOfContents();

    public Book(String title) {
        super(title);
    }

    public Author addAuthor(String name) {
        Author a = new Author(name);
        authors.add(a);
        return a;
    }

    public void addAuthor(Author author) {
        authors.add(author);
    }

    public void addContent(Element el) {
        super.add(el);
    }

    public Section addSection(String name) {
        Section s = new Section(name);
        super.add(s);
        tableOfContents.addEntry(name);
        return s;
    }

    public List<Author> getAuthors() {
        return authors;
    }

    public TableOfContents getTableOfContents() {
        return tableOfContents;
    }

    @Override
    protected void printThisBefore() {
    }

    @Override
    public void print() {
        System.out.println("Book: " + getTitle());
        System.out.println();
        if (!authors.isEmpty()) {
            System.out.println("Authors:");
            for (Author a : authors) {
                a.print();
            }
            System.out.println();
        }
        super.print();
    }
}
